pkg/serviceaccount/jwt_external.go

```diff
@@ -0,0 +1,174 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package serviceaccount
+
+import (
+	"context"
+	"fmt"
+
+	"google.golang.org/grpc"
+	jose "gopkg.in/square/go-jose.v2"
+	"gopkg.in/square/go-jose.v2/jwt"
+
+	"k8s.io/apiserver/pkg/authentication/authenticator"
+	certutil "k8s.io/client-go/util/cert"
+	"k8s.io/klog"
+	externalsigner "k8s.io/kubernetes/pkg/serviceaccount/externalsigner/v1beta1"
+)
+
+func ExternalJWTTokenGenerator(iss string, socketPath string) (TokenGenerator, error) {
```

Comment:
Would the diff be smaller if we just implement an OpaqueSigner and add an indirection to the Key, e.g. `GetKeys() []jose.JSONWebKey` that get's called every verification?

Reply:
The keyID  is required when instantiating  `jose.NewSigner()`, which happens in `JWTTokenGenerator()`. I figured it would be a more readable diff this way, but open to suggestions.